src/service/impl/bookmarkServiceImpl.ts

```diff
@@ -0,0 +1,105 @@
+import { UserInfo } from '../../model/jwt';
+import { PaginationArg } from '../../model/paginationArg';
+import { paginate } from '../../util/paginate';
+import { BookmarkService } from '../bookmarkService';
+
+@Resolver()
+export class BookmarkServiceImpl implements BookmarkService {
+
+    @Mutation(() => Boolean)
+    @Authorized("realm:skiosa-user")
+    async addBookmark(@CurrentUser() currentUserInfo: UserInfo, @Arg("articleId") articleId: number): Promise<boolean> {
+        const userRepository = dataSource.getRepository(User);
+        const articleRepository = dataSource.getRepository(Article);
+
+        const currentUser = await getCurrentUser(userRepository, currentUserInfo);
+
+        return new Promise((resolve, reject) => {
+            articleRepository.findOneBy({
+                id: articleId,
+            }).then(article => {
+                if (article) {
+                    currentUser.bookmarks?.push(article);
+                    userRepository.manager.save(currentUser);
+                    resolve(true);
+                } else {
+                    reject(new Error("Article does not exist"));
+                }
+            }).catch(() => {
+                reject(new Error("Article does not exist"));
+            });
+        });
+    }
+
+    @Mutation(() => Boolean)
+    @Authorized("realm:skiosa-user")
+    async deleteBookmark(@CurrentUser() currentUserInfo: UserInfo, @Arg("articleId") articleId: number): Promise<boolean> {
+        const userRepository = dataSource.getRepository(User);
+        const articleRepository = dataSource.getRepository(Article);
+
+        const currentUser = await getCurrentUser(userRepository, currentUserInfo);
+
+        return new Promise((resolve, reject) => {
+            articleRepository.findOneBy({
+                id: articleId,
+            }).then(article => {
+                if (article) {
+                    currentUser.bookmarks = currentUser.bookmarks?.filter((bookmark: Article) => {
+                        return bookmark.id !== articleId;
+                    });
+
+                    userRepository.manager.save(currentUser);
+                    resolve(true);
+                } else {
+                    reject(new Error("Article does not exist"));
+                }
+            }).catch(() => {
+                reject(new Error("Article does not exist"));
+            });
+        });
+    }
+
+    @Query(() => [Article])
+    @Authorized("realm:skiosa-user")
+    async bookmarks(@CurrentUser() currentUserInfo: UserInfo, @Arg("PaginationArg", { nullable: true }) paginated?: PaginationArg): Promise<Article[]> {
+        const userRepository = dataSource.getRepository(User);
+        const currentUser = await getCurrentUser(userRepository, currentUserInfo);
+
+        if (!currentUser.bookmarks || currentUser.bookmarks.length === 0) {
+            return [];
+        }
+
+        return paginate(currentUser.bookmarks, paginated);
+    }
```

Comment:
getCurrentUser can throw errors. Awaiting a Promise which can reject is unclean please use .then()

Reply:
should be fixed